state-commitments/optimized/history_commitment.go

```diff
@@ -0,0 +1,461 @@
+				return common.Hash{}, err
+			}
+			if _, err := h.keccak.Write(leaves[2*i+1][:]); err != nil {
+				return common.Hash{}, err
+			}
+			if _, err := h.keccak.Read(leaves[i][:]); err != nil {
+				return common.Hash{}, err
+			}
+			h.keccak.Reset()
+		}
+		if m&1 == 1 {
+			// Check to ensure m-1 is a valid index.
+			if _, err := h.keccak.Write(leaves[m-1][:]); err != nil {
+				return common.Hash{}, err
+			}
+			if j < len(fillers) { // Check to prevent index out of range for fillers.
+				if _, err := h.keccak.Write(fillers[j][:]); err != nil {
+					return common.Hash{}, err
+				}
+			} else {
+				// Handle the case where j is out of range for fillers.
+				return common.Hash{}, errors.New("insufficient fillers")
+			}
+			if _, err := h.keccak.Read(leaves[(m-1)/2][:]); err != nil {
+				return common.Hash{}, err
+			}
+			h.keccak.Reset()
+		}
+		m = (m + 1) / 2
+	}
+	return leaves[0], nil
+}
+
+// computeVirtualSparseTree returns the htr of a hashtree where the first layer
+// is passed as leaves, the completed with the last leaf until it reaches
+// virtual and finally completed with zero hashes until it reaches limit.
+// limit is assumed to be either 0 or a power of 2 which is greater or equal to
+// virtual. If limit is zero it behaves as if it were the smallest power of two
+// that is greater or equal than virtual.
+//
+// The algorithm is split in three different logic parts:
+//
+//  1. If the virtual length is less than half the limit (this can never happen
+//     in the first iteration of the algorithm), then the first half of the tree
+//     is computed by recursion and the second half is a zero hash of a given
+//     depth.
+//  2. If the leaves all fit in the first half, then we can optimize the first
+//     half to being a simple sparse tree, just that instead of filling with zero
+//     hashes we fill with the precomputed virtual hashes. This is the most common
+//     starting scenario. The second part is computed by recursion.
+//  3. If the leaves do not fit in the first half, then we can compute the first half of
+//     the tree as a normal full hashtree. The second part is computed by recursion.
+func (h *HistoryCommitter) computeVirtualSparseTree(leaves []common.Hash, virtual, limit uint64) (common.Hash, error) {
+	m := uint64(len(leaves))
+	if m == 0 {
+		return common.Hash{}, errors.New("nil leaves")
+	}
+	if virtual < m {
+		return common.Hash{}, fmt.Errorf("virtual %d should be >= num leaves %d", virtual, m)
+	}
+	var err error
+	if limit == 0 {
+		limit = nextPowerOf2(virtual)
+		n := 1
+		if virtual > m {
+			logValue := math.Log2(float64(limit))
+			n = int(logValue) + 1
+		}
+		h.lastLeafFillers, err = h.precomputeRepeatedHashes(&leaves[m-1], n)
+		if err != nil {
+			return common.Hash{}, err
+		}
+	}
+	if limit == 1 {
+		return leaves[0], nil
+	}
+	if limit < virtual {
+		return common.Hash{}, fmt.Errorf("limit %d should be >= virtual %d", limit, virtual)
+	}
```

Comment:
This should probably be moved before the return above if `limit == 1`, reason being that if you pass `limit = 1` but `virtual = 5` this is most certainly an error.

Reply:
Done.